### PR TITLE
pageserver: reuse idempotency keys across metrics sinks

### LINKED ISSUE #8605: Pageserver: billing events sent to Vector and S3 should use the same idempotency key
## Problem

The idempotency key is used to make sure that events are not recorded more than once, although they may be submitted multiple times in some cases to make sure they are not lost.

We should calculate the idempotency key beforehand and send the same events to S3 and Vector.

Relates (internal issue): https://github.com/neondatabase/cloud/issues/9824

## Detail

Consumption metrics for billing are written over a socket to an external service (Vector), and also written to S3 for posterity.

In `consumption_metrics.rs`, we call two output methods with the same vector of metric values:
- `upload::upload_metrics_bucket`
- `upload::upload_metrics_http`

Each of these ultimately uses RawMetric::as_event on each metric to add an "idempotency key" to the entry: this enables the billing system to receive delta metrics (e.g. data written since last sample) without risking double-counting on retries.

To ensure the S3 output and the Vector output have the same idempotency key, we need to pull the calculation of the keys up into `collect_metrics`, and pass those with the RawMetrics into each upload function, so that the uploads aren't independently calculating different keys.

### PR DESCRIPTION
**Pull request by erikgrinaker**
## Problem

Metrics event idempotency keys differ across S3 and Vector. The events should be identical.

Resolves #8605.

## Summary of changes

Pre-generate the idempotency keys and pass the same set into both metrics sinks.

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

**Comment by erikgrinaker** [comment]
cc @jcsp

There's a minor hazard here: the idempotency keys use the current time and a random number. Since we're generating these in a tight loop, it's possible (but unlikely) to generate colliding keys. This is less of a risk on Linux, which uses nanosecond clock precision. We could consider using e.g. UUIDv6 to generate guaranteed unique keys, but I'll leave that for a follow-up PR if desired.

**Comment by jcsp** [comment]
> Since we're generating these in a tight loop, it's possible (but unlikely) to generate colliding keys.

That should be okay: the idempotency keys need to be unique within the metrics we output for a particular tuple of metric name & parameters (e.g. size of a particular tenant), but it's fine if two tenants end up outputting metrics with the same idempotency key.

**Review by jcsp — APPROVED** [review]
LGTM: it could be a little more efficient to put the keys into the same Vec where we store the RawMetric, but on  balance I prefer the approach here as it maintains the separation of concerns between the code that does metric collection and the code that knows about transmission and idempotency keys.

**Comment by erikgrinaker** [comment]
TFTR!

> the idempotency keys need to be unique within the metrics we output for a particular tuple of metric name & parameters

I see -- that's fine then.

> on balance I prefer the approach here as it maintains the separation of concerns between the code that does metric collection and the code that knows about transmission and idempotency keys

Yeah, that was my thought as well.

File: pageserver/src/consumption_metrics/upload.rs [hunk 1]
```diff
@@ -268,7 +269,7 @@ impl RawMetricExt for RawMetric {
     }
 }
 
-trait KeyGen<'a>: Copy {
+pub(crate) trait KeyGen<'a> {
     fn generate(&self) -> IdempotencyKey<'a>;
 }
```
Comment (koivunej) [inline pageserver/src/consumption_metrics/upload.rs:274]: Is this used from somewhere now?

File: pageserver/src/consumption_metrics.rs [hunk 2]
```diff
@@ -143,6 +145,12 @@ async fn collect_metrics(
         // these are point in time, with variable "now"
         let metrics = metrics::collect_all_metrics(&tenant_manager, &cached_metrics, &ctx).await;
 
+        // Pre-generate event idempotency keys, to reuse them across the bucket
+        // and HTTP sinks.
+        let idempotency_keys = std::iter::repeat_with(|| node_id.as_str().generate())
+            .take(metrics.len())
+            .collect_vec();
```
Comment (koivunej) [inline pageserver/src/consumption_metrics.rs:152]: Doesn't this generate N idempotency keys out of the same string? (basically cloning a string N times?)

File: pageserver/src/consumption_metrics/upload.rs (hunk 1 above)
Comment (erikgrinaker) [reply pageserver/src/consumption_metrics/upload.rs:274]: Discussed on Slack, I'll submit a PR tomorrow to remove this trait and just construct the idempotency key directly.

File: pageserver/src/consumption_metrics.rs (hunk 2 above)
Comment (erikgrinaker) [reply pageserver/src/consumption_metrics.rs:152]: It generates a unique idempotency key for each iteration, based on the same node ID.

**Pull request merged**